pkg/skaffold/docker/debugger/transform.go

```diff
@@ -0,0 +1,111 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package debugger
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/volume"
+	"github.com/sirupsen/logrus"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/debug"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/debug/types"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/graph"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+var (
+	shouldTransform    bool
+	SupportVolumeMount = volume.VolumeCreateBody{Name: debug.DebuggingSupportFilesVolume}
+	TransformImage     = transformImage // For testing
+)
+
+type Transform func(string) string
+
+func EnableTransforms() {
+	shouldTransform = true
+}
+
+func transformImage(ctx context.Context, artifact graph.Artifact, cfg *container.Config, insecureRegistries map[string]bool, debugHelpersRegistry string) (map[string]types.ContainerDebugConfiguration, []*container.Config, error) {
+	if !shouldTransform {
+		return nil, nil, nil
+	}
+	portAvailable := func(port int32) bool {
+		return isPortAvailable(cfg, port)
+	}
+	portAlloc := func(desiredPort int32) int32 {
+		return util.AllocatePort(portAvailable, desiredPort)
+	}
+
+	adapter := NewAdapter(cfg)
+	imageConfig, err := debug.RetrieveImageConfiguration(ctx, &artifact, insecureRegistries)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	configurations := make(map[string]types.ContainerDebugConfiguration)
+	var initContainers []*container.Config
+	// the container images that require debugging support files
+	var containersRequiringSupport []*container.Config
+	// the set of image IDs required to provide debugging support files
+	requiredSupportImages := make(map[string]bool)
+
+	if configuration, requiredImage, err := debug.TransformContainer(adapter, imageConfig, portAlloc); err == nil {
+		configurations[adapter.GetContainer().Name] = configuration
+		if len(requiredImage) > 0 {
+			logrus.Infof("%q requires debugging support image %q", cfg.Image, requiredImage)
+			containersRequiringSupport = append(containersRequiringSupport, cfg)
+			requiredSupportImages[requiredImage] = true
+		}
+	} else {
+		logrus.Warnf("Image %q not configured for debugging: %v", cfg.Image, err)
+	}
+
+	// check if we have any images requiring additional debugging support files
+	if len(containersRequiringSupport) > 0 {
+		logrus.Infof("Configuring installation of debugging support files")
+		// the initContainers are responsible for populating the contents of `/dbg`
+		for imageID := range requiredSupportImages {
+			supportFilesInitContainer := &container.Config{
+				Image:   fmt.Sprintf("%s/%s", debugHelpersRegistry, imageID),
+				Volumes: map[string]struct{}{"/dbg": {}},
+			}
+			initContainers = append(initContainers, supportFilesInitContainer)
+		}
+		// the populated volume is then mounted in the containers at `/dbg` too
```

Comment:
Mark this as a TODO, and we should modify the container-debug-support images to specify this location as a volume.

Reply:
this part will always be necessary, do you mean put a TODO for declaring `VOLUME /dbg` in the container-debug-support images?